### ISSUE TITLE
support rules specifying origin and/or destination URL path(s)

### ISSUE THREAD

**Issue by msxfm** [RequestPolicyContinued/requestpolicy#299]
<a href="https://github.com/alexlehm"><img src="https://avatars.githubusercontent.com/u/156144?" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [alexlehm](https://github.com/alexlehm)**
_Monday Apr 09, 2012 at 20:40 GMT_
_Originally opened as https://github.com/RequestPolicy/requestpolicy/issues/299_

---

It would be helpful if whitelist entries could contain urls with subdirs in addition to the domain, e.g. for the url redirect on search engines or on facebook this would be useful, since I wouldn't want to whitelist the complete domain but all urls that are redirected away from the domain, e.g.

on google.com, when I click on an url in the search results, the url is first redirected to www.google.com/url?...url=... which then in turn redirects to the target url.

The whitelist entry for this would have to be allow origin google.com/url\?.*

**Comment by mk-pmb** [comment]
When adding regexps, let's find a clever way to deal with dots and slashes, as lots of backslashes impede reading. Third party tools can help with rule creation but less with not debugging.

**Comment by myrdd** [comment]
@mk-pmb Justin aimed at having both "path prefix" and "path regex". If a "path prefix" is specified, the path has to start with the given string. For more complex cases you need to use regex. For performance reasons I'd approach JavaScript's built-in regex syntax, which indeed requires to escape dots. However, when you use the RegExp _constructor_, which is what I'll do, you don't need to escape slashes. See:

```javascript
r = new RegExp("/");
// …is equivalent to…
r = /\//;
```

I think that's good enough.

---

Implementation note: care must be taken with backslashes in combination with JSON. See:

```javascript
let s = "\\."; // string will be "\."
let r = new RegExp(s); // regexp will be /\./
let j = JSON.stringify({pathRegex: s}); // JSON string will be "{"pathRegex":"\\."}"
```

**Comment by mk-pmb** [comment]
:+1:

**Comment by alexlehm** [comment]
I like the way this is done in Foxyproxy, you can either use wildcard patterns or regexp, for most use cases a wildcard is sufficient like "*://*.yahoo.com/*" and for more complex ones you can use a regexp

**Comment by mk-pmb** [comment]
However, there are also quite some cases where users mistakenly think that wildcards are safe enough. Re-using the example `*://*.yahoo.com/*`, some people might be surprised when their Yahoo proxy setting is used for http://mallory.example.net/tracker.php?from=http://mail.yahoo.com/whatever&spam=XOBiLiZLlnxr .

**Comment by myrdd** [comment]
Foxyproxy's URL wildcards are nice, but the way RP works it would be _path_ wildcards. Not sure if it is very helpful; Foxyproxy simply translates the wildcard rules to regular expressions. I think we're safe only supporting path-regex (and path-prefix). Instead of writing `/some/*/path/*` you need to write `^/some/.*/path/` (or `^/some/[^/]*/path/`), which is ok IMHO. If desired, path-wildcards can be added to a later time.

Btw @alexlehm I've edited your post due to a parsing error in the wildcard string. Asterisk (`*`) means "bold" in Markdown. Surrounding with backticks (`` ` ``) helps.

**Comment by alexlehm** [comment]
@mk-pmb yes you're right, I had that with some tool that tried to match *play.google.com* and matched a search parameter where the search text was play.google.com

**Comment by alexlehm** [comment]
actually the wildcard problem applies to regexp patterns as well, since a "naive" user would use `.*://*.yahoo.com/.*` which would match the domain somewhere in the path or query. Correct would be something like `[^:]+://[^/]+.yahoo.com/.*`

**Comment by myrdd** [comment]
@alexlehm yes, so the usage of `[^/]` in the path-regex should be encouraged.

**Comment by mk-pmb** [comment]
I also suggest checking whether the matched portion of the string is exactly equal to the original string, so we don't have to worry what exactly can be matched by `^` and `$` and what happens when URLs grow really huge. Just some days ago my `sed` failed due to maximum line length, and I think that could be deep down on C layers, as I think sed rather provides a thin wrapper than doing a lot of regexp magic itself. If so, we shouldn't be surprised if a browser is affected at some future time.

**Comment by myrdd** [comment]
Please stay on-topic. This issue is only about _path_ RegExps. RPC handles the URI/URL parts "scheme", "host", and "path" completely separately. Having a rule with a "Path-RegExp" won't affect host-matching _at all_.

By the way, the wildcards allowed in RPC's domain-name-specification do _not_ translate into regexps. The "host" string is split by the literal dots (`.`).

> I also suggest checking whether the matched portion of the string is exactly equal to the original string, so we don't have to worry what exactly can be matched by `^` and `$` and what happens when URLs grow really huge.

Could you please give some further explanation what you mean? Which two strings do you want to compare? What do you want to prevent by doing so?

Regarding your comparison with sed, JavaScript has no such maximum string length. Since there are multiple add-ons operating on URLs using RegExp, we should be safe here.

What's exactly the problem with non-alphanumeric characters? What kind of malicious interpretation?

**Comment by mk-pmb** [comment]
> This issue is only about *path* RegExps.

Right, I'm sorry, I forgot that and for some reason, at time of my posting, your reminder and some other messages weren't visible to me yet. All attack vectors I had in mind would require a full URL with a hostname, so they don't apply here.
